### PR TITLE
`Communication`: Allow to navigate to channel related pages directly from the messages tab

### PR DESCRIPTION
**Pull request by lennart-keller**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This PR is supposed to make it easier to go to the exercise/lecture/exam that belongs to the according channel.

### Description

A new button is added to the menu next to a lecture/exercise/exam channel, that allows to quickly navigate to the according lecture/exercise/exam page.
The server sends the related exercise/exam/lecture id to the client and depending on the subtype of the channel, the client navigates to the according page when the button is clicked.

### Steps for Testing

Prerequisites:
- 1 Course with exercise/lecture/exam channels
- 1 Student

1. Navigate to the messages tab
2. Click the menu button next to a lecture/exercise/exam channel and select "Go to ..."/"Gehe zu..." and check that you are visiting the correct lecture/exercise/exam page
3. When selecting a lecture/exercise/exam channel, there should also be a navigation button at the top right (see screenshots)
4. Check that the buttons are labeled correctly for all 3 types
5. Check that there are no "Go to..." buttons for general channels or group chats / direct messages
6. Check that you can also open the exercise/lecture/exam page in a new tab by right clicking the two new navigation buttons

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [channel.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5070-7/artifact/shared/Coverage-Report-Client-Tests/app/entities/metis/conversation/channel.model.ts.html) | 93.75% | ✅ |
| [conversation-header.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5070-7/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/layout/conversation-header/conversation-header.component.ts.html) | 96.07% | ✅ |
| [conversation-sidebar-entry.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5070-7/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-sidebar-section/conversation-sidebar-entry/conversation-sidebar-entry.component.ts.html) | 93.22% | ✅ |
| [conversation.util.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5070-7/artifact/shared/Coverage-Report-Client-Tests/app/overview/course-conversations/other/conversation.util.ts.html) | 94.73% | ✅ |
| [metis.service.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5070-7/artifact/shared/Coverage-Report-Client-Tests/app/shared/metis/metis.service.ts.html) | 89.83% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ChannelDTO.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5070-7/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis.conversation.dtos/ChannelDTO.html) | 100% | ✅ |

### Screenshots

New navigation button within channel option menu:
option menu

New navigation button within conversation header:
conversation header

### CODE REVIEW

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
Could you please add the option for right clicks on your new buttons?
Additionally check again if closed test exams work properly, i.e. you cannot open them

File: src/main/webapp/app/shared/metis/metis.service.ts [hunk 1]
```diff
@@ -368,6 +369,39 @@ export class MetisService implements OnDestroy {
         return `/courses/${this.getCourse().id}/lectures/${lectureId}`;
     }
 
+    /**
+     * returns the router link required for navigating to the exam
+     * @param {string} lectureId ID of the exam to be navigated to
```
Comment (rriyaldhi) [inline src/main/webapp/app/shared/metis/metis.service.ts:374]:
```suggestion
     * @param {string} exam ID of the exam to be navigated to
```
I think this should be exam ID

File: src/test/javascript/spec/component/overview/course-conversations/layout/conversation-header/conversation-header.component.spec.ts [hunk 2]
```diff
@@ -35,8 +51,18 @@ examples.forEach((activeConversation) => {
                     MockComponent(GroupChatIconComponent),
                     MockComponent(FaIconComponent),
                     MockPipe(ArtemisTranslatePipe),
+                    RouterTestingModule.withRoutes([
+                        { path: 'courses/:courseId/lectures/:lectureId', component: CourseLectureDetailsComponent },
+                        { path: 'courses/:courseId/exercises/:lectureId', component: CourseExerciseDetailsComponent },
```
Comment (rriyaldhi) [inline src/test/javascript/spec/component/overview/course-conversations/layout/conversation-header/conversation-header.component.spec.ts:56]: Is this intentional to have :lectureId here?

File: src/test/javascript/spec/component/overview/course-conversations/layout/conversation-header/conversation-header.component.spec.ts [hunk 3]
```diff
@@ -35,8 +51,18 @@ examples.forEach((activeConversation) => {
...
+                        { path: 'courses/:courseId/exams/:lectureId', component: ExamDetailComponent },
```
Comment (rriyaldhi) [inline src/test/javascript/spec/component/overview/course-conversations/layout/conversation-header/conversation-header.component.spec.ts:57]: Is this intentional to have :lectureId here?

**Review by rriyaldhi — COMMENTED** [review]
Looks good overall, but left some questions.

File: src/main/webapp/app/shared/metis/metis.service.ts (hunk 1 above)
Comment (lennart-keller) [reply src/main/webapp/app/shared/metis/metis.service.ts:374]: Yes, you are right :D

File: src/test/javascript/spec/component/overview/course-conversations/layout/conversation-header/conversation-header.component.spec.ts (hunk 3 above)
Comment (lennart-keller) [reply src/test/javascript/spec/component/overview/course-conversations/layout/conversation-header/conversation-header.component.spec.ts:57]: No (id doesn't really matter, but it is inconsistent, you are right). I corrected it

File: src/test/javascript/spec/component/overview/course-conversations/layout/conversation-header/conversation-header.component.spec.ts (hunk 2 above)
Comment (lennart-keller) [reply src/test/javascript/spec/component/overview/course-conversations/layout/conversation-header/conversation-header.component.spec.ts:56]: No ;)

**Review by rriyaldhi — DISMISSED** [review]
Code looks good.

File: src/main/webapp/app/overview/course-conversations/layout/conversation-header/conversation-header.component.html [hunk 4]
```diff
@@ -21,6 +21,13 @@ <h3 class="conversation-name d-inline-block rounded p-2 info" (click)="openConve
         <div>
             <div class="btn-toolbar p-2" role="toolbar">
                 <div class="btn-group me-2" role="group">
+                    <a
+                        *ngIf="getAsChannel(activeConversation)?.subTypeReferenceId"
+                        class="btn btn-outline-secondary sub-type-reference"
+                        [routerLink]="metisService.getLinkForChannelSubType(getAsChannel(activeConversation))"
```
Comment (Strohgelaender) [inline src/main/webapp/app/overview/course-conversations/layout/conversation-header/conversation-header.component.html:27]: can you create helper variables for these? Using methods here calls them over and over again even though these are constants.

File: src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-sidebar-section/conversation-sidebar-entry/conversation-sidebar-entry.component.html [hunk 5]
```diff
@@ -21,6 +21,14 @@
...
             </button>
             <div ngbDropdownMenu>
+                <a
+                    *ngIf="getAsChannel(conversation)?.subTypeReferenceId"
+                    class="sub-type-reference"
+                    ngbDropdownItem
+                    [routerLink]="[metisService.getLinkForChannelSubType(getAsChannel(conversation))]"
+                >
+                    {{ getChannelSubTypeReferenceTranslationKey(getAsChannel(conversation)?.subType) | artemisTranslate }}
+                </a>
```
Comment (Strohgelaender) [inline src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-sidebar-section/conversation-sidebar-entry/conversation-sidebar-entry.component.html:31]: same here.

File: src/main/webapp/app/overview/course-conversations/layout/conversation-header/conversation-header.component.html (hunk 4 above)
Comment (lennart-keller) [reply src/main/webapp/app/overview/course-conversations/layout/conversation-header/conversation-header.component.html]: ok, I use helper variables now and also replaced other places where these methods were called in the template before.

File: src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-sidebar-section/conversation-sidebar-entry/conversation-sidebar-entry.component.html (hunk 5 above)
Comment (lennart-keller) [reply src/main/webapp/app/overview/course-conversations/layout/conversation-selection-sidebar/conversation-sidebar-section/conversation-sidebar-entry/conversation-sidebar-entry.component.html]: same here :)

**Review by JohannesStoehr — APPROVED** [review]
Clickable links and code changes look good to me 👍

**Review by Strohgelaender — APPROVED** [review]
looks good

**Review by rriyaldhi — APPROVED** [review]
Reapprove code

**Review by pellumbbaboci — APPROVED** [review]
Code looks good 👍  and tested on ts1 works as expected. 

One additional idea is to navigate from exercise/lecture to dedicated channel as its harder to navigate manually from exercise/lecture to channel.  This of course can be addressed in a separate PR.

**Pull request merged**